### PR TITLE
internal: Tweak change collapsing

### PR DESCRIPTION
**Pull request by lnicola**
CC https://github.com/rust-lang/rust-analyzer/pull/14025#discussion_r1087946555.

### CODE REVIEW

File: crates/rust-analyzer/src/global_state.rs [hunk 1]
```diff
@@ -191,57 +192,39 @@ impl GlobalState {
...
+                    .and_modify(|change| match (change, changed_file.change_kind) {
+                        // latter `Delete` wins
+                        (change, Delete) => *change = Delete,
+                        // merge `Create` with `Create` or `Modify`
+                        (Create, Create | Modify) => {}
+                        // collapse identical `Modify`es
+                        (Modify, Modify) => {}
+                        // equivalent to `Modify`
+                        (change @ Delete, Create) => *change = Modify,
+                        // shouldn't occur, but collapse into `Create`
```
Comment (lnicola) [inline crates/rust-analyzer/src/global_state.rs:213]: Small change in behavior on these two.

File: crates/rust-analyzer/src/global_state.rs [hunk 2]
```diff
@@ -191,57 +192,39 @@ impl GlobalState {
...
-            changed_files.sort_by_key(|file| file.file_id);
-            // We need to fix up the changed events a bit, if we have a create or modify for a file
-            // id that is followed by a delete we actually no longer observe the file text from the
-            // create or modify which may cause problems later on
-            let mut collapsed_create_delete = false;
-            changed_files.dedup_by(|a, b| {
+            // We need to fix up the changed events a bit. If we have a create or modify for a file
+            // id that is followed by a delete we actually skip observing the file text from the
+            // earlier event, to avoid problems later on.
+            file_changes.clear();
```
Comment (Veykril) [inline crates/rust-analyzer/src/global_state.rs:198]: this clear seems unnecessary

**Review by Veykril — APPROVED** [review]

File: crates/rust-analyzer/src/global_state.rs (hunk 2 above)
Comment (lnicola) [reply crates/rust-analyzer/src/global_state.rs]: Indeed, fixed.

**Comment by lnicola** [comment]
This wasn't was actually no longer dropping the `Create`-`Delete` pairs, hope it's better now :grimacing:.

**Pull request merged**